### PR TITLE
remove channels slice and just use map

### PR DESCRIPTION
**Pull request by melekes**
as far as I know, we don't need order while sending a msg to channels
nor when we report a status.

Refs #1119

### CODE REVIEW

**Comment by ebuchman** [comment]
Only concern might be performance of iterating over a map. Can we get a quick benchmark of the difference for say 8, 16, 32, etc. items.

Currently we have about 8 channels I believe. Maybe it will double. I doubt it would be as high as 100 anytime soon but might be good to know

**Comment by melekes** [comment]
```
BenchmarkIterateOverChannelsMap8-2              20000000               114 ns/op
BenchmarkIterateOverChannelsMap16-2              5000000               294 ns/op
BenchmarkIterateOverChannelsMap32-2              3000000               559 ns/op
BenchmarkIterateOverChannelsSlice8-2            200000000                7.71 ns/op
BenchmarkIterateOverChannelsSlice16-2           100000000               14.1 ns/op
BenchmarkIterateOverChannelsSlice32-2            3000000               560 ns/op
```

> Only concern might be performance of iterating over a map

why is this a concern at all if we use map when we send something to a channel???
```
	// Send message to channel.
	channel, ok := c.channels[chID]
```

i.e. we only use slice when reporting status and updating stats

**Comment by ebuchman** [comment]
> why is this a concern at all if we use map when we send something to a channel???

It's just a lookup, we're not iterating over the whole map.

Looks like small slices are ~20x faster to iterate through

File: p2p/conn/connection.go [hunk 1]
```diff
@@ -524,14 +518,16 @@ func (c *MConnection) Status() ConnectionStatus {
 	status.SendMonitor = c.sendMonitor.Status()
 	status.RecvMonitor = c.recvMonitor.Status()
 	status.Channels = make([]ChannelStatus, len(c.channels))
-	for i, channel := range c.channels {
+	i := 0
+	for _, channel := range c.channels {
```
Comment (ebuchman) [inline p2p/conn/connection.go]: it does mean the channels will be in a different order here every time over the RPC. Could be weird.

Comment (melekes) [reply p2p/conn/connection.go]: I would change RPC to return a map instead. I.e. if somebody relies on order here, then he/she is obviously doing wrong thing

Comment (melekes) [reply p2p/conn/connection.go]: Like https://github.com/tendermint/tendermint/pull/1141/commits/da59c46a5a0589803aaaebdcb4624788cba430ff

**Comment by ebuchman** [comment]
Why'd you close it ?

> I would change RPC to return a map instead. I.e. if somebody relies on order here, then he/she is obviously doing wrong thing

I don't disagree, but Is this normal behaviour for an API ? Seems inevitable in Go APIs due to explitly randomized iteration but just curious ...

**Pull request closed without merging**